app/src/main/java/protect/card_locker/Utils.java

```diff
@@ -422,16 +423,18 @@ static public String formatBalanceWithoutCurrencySymbol(BigDecimal value, Curren
     static public BigDecimal parseBalance(String value, Currency currency) throws ParseException {
         NumberFormat numberFormat = NumberFormat.getInstance();
 
+        if (numberFormat instanceof DecimalFormat) {
+            ((DecimalFormat) numberFormat).setParseBigDecimal(true);
+        }
+
         if (currency == null) {
             numberFormat.setMaximumFractionDigits(0);
         } else {
             numberFormat.setMinimumFractionDigits(currency.getDefaultFractionDigits());
             numberFormat.setMaximumFractionDigits(currency.getDefaultFractionDigits());
         }
 
-        Log.d(TAG, numberFormat.parse(value).toString());
-
-        return new BigDecimal(numberFormat.parse(value).toString());
+        return (BigDecimal) numberFormat.parse(value);
```

Comment:
Rarely this could fail since `NumberFormat.getInstance()` is not guaranteed to return an instance of `DecimalFormat`
We could add a function to handle those cases
```
private static final double LargestPreciseDouble = (double) (1l << 53);
static{
    assert (LargestPreciseDouble + 1.0) == LargestPreciseDouble;
    assert (LargestPreciseDouble - 1.0) != LargestPreciseDouble;
}

private static BigDecimal fromParsed(Number parsed){
    if(parsed instanceof BigDecimal)
      return (BigDecimal) parsed;

    final double d = parsed.doubleValue();
    if(d >= LargestPreciseDouble)
      return new BigDecimal(parsed.longValue());
    return new BigDecimal(d);
}
```
    
and then make this line `return fromParsed(numberFormat.parse(value));`

Reply:
First of all thank you for your feedback !

I tested the code that you suggested to use and I observed that when I add for example the number 9999999999999999999 (19 times the number 9 or more) it is converted to 9,223,372,036,854,775,807 which is the value of the `LargestPreciseDouble` variable. 
You mention with the assertion that  `(LargestPreciseDouble + 1.0) == LargestPreciseDouble` must hold. 
I believe that when you assign a value to the variable d in the line `final double d = parsed.doubleValue();` if `parsed` is greater than `LargestPreciseDouble` , then d is assigned with the value of `LargestPreciseDouble` and that's why this behavior is observed. Do you have any suggestions how we could overcome this ?

Reply:
Hmm, interesting. I tried @electroCutie's patch and can't reproduce your issue (with 20 taps of 9):

[Screen_recording_20240421_132234.webm](https://github.com/CatimaLoyalty/Android/assets/1885159/85f795ea-6c20-4e34-8c74-33ca4574fe05)

Are you testing it the same way as me? Entering numbers from the keyboard and then tapping away? Or are you doing something else?

Reply:
Well I do the same steps and this is what happens. 

[Uploading Catima_Loyalty_20240421_213602.webm…]()

Reply:
Well I do the same steps but this is what happens 
[Catima_Loyalty_20240421_213602.webm](https://github.com/CatimaLoyalty/Android/assets/115111582/161ce7b8-057e-4685-9252-3c88838f7bae)

Reply:
Hmm, interesting. I can't reproduce it. Could you please push the code with @electroCutie's change so we can be sure we're talking about the same code?

Reply:
I pushed the code with @electroCutie's changes but then noticed that mistakenly I had erased these lines 
`if (numberFormat instanceof DecimalFormat) {`
            `(DecimalFormat)` `numberFormat).setParseBigDecimal(true);`
  `}`

By adding them it worked as expected for me as well !!